app/models/main-graph.ts

```diff
@@ -0,0 +1,336 @@
+ * the backer object (core/drawable) references another core or drawable
+ * this will return a `BridgingProxy` in its place. Sets will intellegently
+ * sync the backends even if the set `v` is a `BridgingProxy`
+ */
+export class BridgingProxy {
+    proxy: { [a: string]: any };
+
+    constructor(public core: CoreElement, public drawable: Drawable, public graph: MainGraph) {
+        this.proxy = new Proxy(core.data, {
+            set: (data, k, v) => this.set(k, v),
+            get: (data, k) => this.get(k),
+        });
+    }
+
+    set(k: PropertyKey, v: any, updateDrawable = true) {
+        let drawableValue = v;
+        let coreValue = v;
+
+        // If v is a drawable element or a core element,
+        // then when it gets set in the various proxied data
+        // structures, set it differently. 
+
+        if (v instanceof BridgingProxy) {
+            drawableValue = v.drawable;
+            coreValue = v.core;
+        } else if (v instanceof CoreElement) {
+            drawableValue = drawableFromAny(v, this.graph.bridges);
+        } else if (v instanceof Drawable) {
+            coreValue = coreFromAny(v, this.graph.bridges);
+        }
+
+        if (updateDrawable && Object.keys(this.drawable).indexOf(k.toString()) !== -1) {
+            (this.drawable as any)[k] = drawableValue;
+        }
+
+        this.graph.changed.emit(new UndoableChange(this, k, this.core.data[k], coreValue));
+
+        this.core.data[k] = coreValue;
+        return true;
+    }
+
+    get(k: PropertyKey) {
+        const coreValue = this.core.data[k];
+        // if this is something that should be proxied
+        const proxiedValue = this.graph.bridges.getB(coreValue);
+        // return the proxy. That is, any changes made to the returned 
+        // object should propogate to both backers
+        return proxiedValue !== undefined ? proxiedValue : coreValue;
+    }
+}
+
+function drawableFromAny(a: any, bridges: DoubleMap<Drawable, CoreElement, BridgingProxy>) {
+    if (a instanceof CoreElement) {
+        const bridge = bridges.getB(a);
+        if (bridge === undefined) {
+            throw "trying to reference a core element that doesn't have an analogous drawable";
+        }
+        a = bridge.drawable;
+    }
+    return a;
+}
+
+function coreFromAny(a: any, bridges: DoubleMap<Drawable, CoreElement, BridgingProxy>) {
+    if (a instanceof Drawable) {
+        const bridge = bridges.getA(a);
+        if (bridge === undefined) {
+            throw "trying to reference a drawable element that doesn't have an analogous core";
+        }
+        a = bridge.core;
+    }
+    return a;
+}
+
+export type UndoableEvent = UndoableChange | UndoableAdd | UndoableDelete;
+
+export class UndoableAdd {
+    constructor(public bridge: BridgingProxy) { }
+}
+
+export class UndoableDelete {
+    constructor(public bridge: BridgingProxy) { }
+}
+
+// TODO: changes can also be adds or deletes
+export class UndoableChange {
```

Comment:
I agree with Sheyne. The changes are useful beyond Undo/Redo. The current implementation is the most workable IMO.